File: src/watch/snapshot.ts
/**
 * Builds the routine snapshot the phone sends to the watch.
 *
 * Pure, so the shape and the limits are unit tested without a device. Values are clamped to
 * `ITEM_BOUNDS` and the lists cut to `IMPORT_LIMITS` here, on the way out, because the watch
 * rejects a snapshot that breaks them (Stability rule 3): one out-of-range row written by an
 * older build must not cost the user every routine on their wrist.
 */
import type { Routine } from '@/domain/types';
import { IMPORT_LIMITS, ITEM_BOUNDS } from '@/transfer/format';
import type { UnitSystem } from '@/utils/format';
import { WATCH_FORMAT_VERSION, WATCH_ROUTINES_FORMAT, type WatchRoutinesDocument } from './format';

const clamp = (value: number, range: { min: number; max: number }) =>
  Number.isFinite(value) ? Math.min(range.max, Math.max(range.min, value)) : range.min;

export function buildWatchRoutines(
  routines: readonly Routine[],
  unitSystem: UnitSystem,
  exportedAt: Date,
): WatchRoutinesDocument {
  return {
    format: WATCH_ROUTINES_FORMAT,
    version: WATCH_FORMAT_VERSION,
    exportedAt: exportedAt.toISOString(),
    unitSystem,
    routines: routines.slice(0, IMPORT_LIMITS.routines).map((routine) => ({
      id: routine.id,
      name: routine.name,
      items: routine.items.slice(0, IMPORT_LIMITS.itemsPerRoutine).map((item) => ({
        id: item.id,
        exerciseId: item.exerciseId,
        exerciseName: item.exerciseName,
        sets: Math.round(clamp(item.sets, ITEM_BOUNDS.sets)),
        reps: item.reps.slice(0, ITEM_BOUNDS.repsLength),
        weightKg: clamp(item.weightKg, ITEM_BOUNDS.weightKg),
        restSeconds: Math.round(clamp(item.restSeconds, ITEM_BOUNDS.restSeconds)),
        notes: item.notes === null ? null : item.notes.slice(0, ITEM_BOUNDS.notesLength),
      })),
    })),
  };
}

File: targets/watch/Core/Snapshot.swift
import Foundation

/// `kinetiq.watch-routines` v1, phone to watch (issue #27). Mirrors `src/watch/format.ts`.
public struct RoutinesSnapshot: Codable, Equatable, Sendable {
    public static let format = "kinetiq.watch-routines"
    public static let version = 1

    public let format: String
    public let version: Int
    public let exportedAt: String
    public let unitSystem: UnitSystem
    public let routines: [Routine]

    public init(format: String, version: Int, exportedAt: String, unitSystem: UnitSystem, routines: [Routine]) {
        self.format = format
        self.version = version
        self.exportedAt = exportedAt
        self.unitSystem = unitSystem
        self.routines = routines
    }
}

public enum UnitSystem: String, Codable, Sendable {
    case metric
    case imperial
}

public struct Routine: Codable, Equatable, Identifiable, Sendable {
    public let id: String
    public let name: String
    public let items: [RoutineItem]

    public init(id: String, name: String, items: [RoutineItem]) {
        self.id = id
        self.name = name
        self.items = items
    }
}

public struct RoutineItem: Codable, Equatable, Identifiable, Sendable {
    public let id: String
    public let exerciseId: String
    public let exerciseName: String
    public let sets: Int
    /// A number or a range, "8" or "8-12". `Reps.target` reads the number a set starts at.
    public let reps: String
    public let weightKg: Double
    public let restSeconds: Int
    public let notes: String?

    public init(
        id: String,
        exerciseId: String,
        exerciseName: String,
        sets: Int,
        reps: String,
        weightKg: Double,
        restSeconds: Int,
        notes: String?
    ) {
        self.id = id
        self.exerciseId = exerciseId
        self.exerciseName = exerciseName
        self.sets = sets
        self.reps = reps
        self.weightKg = weightKg
        self.restSeconds = restSeconds
        self.notes = notes
    }
}
